### PR TITLE
Improvements to BigBlueButton plugin

### PR DESCRIPTION
**Pull request by jfederico**
Closed PR #963 and opened this one

>This PR includes a number of improvements to the integration with BigBlueButton through the plugin although it also includes some additions to the conference module itself.
>
>As discussed with Chris Ward, from the UI perspective the enhancements can be seen in three shot videos we already shared with him (we can send you guys the links through slack).

### CODE REVIEW

**Comment by claydiffrient** [comment]
@jfederico I've pushed this up into our internal code review system.  

One thing I notice missing that we ask for in our [CONTRIBUTING.md](https://github.com/instructure/canvas-lms/blob/stable/CONTRIBUTING.md) is a test plan for the commits.  This helps ensure that our engineers can easily identify how to make sure each commit is doing what it needs to do from a higher level than just reading the code.

**Comment by claydiffrient** [comment]
Our build system failed the commit because of some errors with possible XSS-able things.  Please run the following linters to see what's going on:
`script/xsslint` - XSS linter
`script/rlint` - Ruby linter
`script/eslint` - JavaScript linter

File: app/coffeescripts/views/conferences/ConferenceView.coffee [hunk 1]
```diff
@@ -109,3 +118,139 @@ define [
...
+                class: 'btn btn-small protect icon-lock'
+                text: I18n.t('conferences.recordings.protected', 'Protected')
+                protect: 'true'
+              else if protect=="false"
+                class: 'btn btn-small protect icon-unlock'
+                text: I18n.t('conferences.recordings.unprotected', 'Unprotected')
+                protect: 'false'
+      spinner.hide()
+      if publish
+        $('<a class="'+elem.class+'" data-publish="'+elem.publish+'">'+elem.text+'</a>').insertAfter(spinner)
```
Comment (jenseng) [inline app/coffeescripts/views/conferences/ConferenceView.coffee]: specifically, xsslint complains about the string concatenation here... it doesn't know that `class`/`protect`/`text` are safe, and in fact `text` may be unsafe (e.g. malicious translator)

so this should be something like:

```!coffeescript
$('<a class="'+htmlEscape(elem.class)+'" data-publish="'+htmlEscape(elem.publish)+'">'+htmlEscape(elem.text)+'</a>')
```

note that you'll need to require `str/htmlEscape` at the top of the file so that it's available

Comment (jfederico) [inline app/coffeescripts/views/conferences/ConferenceView.coffee]: Thanks @jenseng I'll take a look

Comment (jfederico) [inline app/coffeescripts/views/conferences/ConferenceView.coffee]: Hi jensen, a couple of questions. 

Fist, why would we need to escape the class and the data if they are hardcoded a few lines up? #L209 and #L211

Second, I can not get to work the require. What else would I need to do? I did this:
```
define [
  'i18n!conferences'
  'jquery'
  'Backbone'
  'jst/conferences/newConference'
  'jquery.google-analytics'
  'compiled/jquery.rails_flash_notifications'
  'str/htmlEscape'
], (I18n, $, {View}, template, htmlEscape) ->
```

But I am still getting:

ConferenceView.js:303 Uncaught TypeError: htmlEscape is not a function
    at ConcludedConferenceView.ConferenceView.togglePublishOrProtectButton (ConferenceView.js:303)

Comment (jenseng) [inline app/coffeescripts/views/conferences/ConferenceView.coffee]: So `xsslint` is not particularly smart, it evaluates each expression pretty much in isolation, and it doesn't read the previous lines to see where stuff comes from. For hardcoded stuff like `class` and `data`, an alternative approach would be to name them differently ... `xsslint` assumes anything named `*Html` is safe, so you could rename those properties to `classHtml` and `dataHtml` respectively

As for the require, you meed to put `str/htmlEscape` higher up, right after `jst/conferences/newConference` ... the elements in that array correspond to the arguments immediately following it (`i18n!conferences` -> `I18n`, `jquery` -> `$`, etc.)

File: public/javascripts/big_blue_button.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        break;
+      default:
+        text=""
+    }
+
+    if($(document).has('#recording_option_legend').length>0){
+      $(document).find('#recording_option_legend').remove();
+    }
+
+    $('<small class="help-text" id="recording_option_legend"><br>'+text+'</small>').insertAfter(element)
```
Comment (simonista) [inline public/javascripts/big_blue_button.js]: one more xsslint error on this line: `[xsslint] possibly XSS-able HTML string concatenation`

File: app/models/big_blue_button_conference.rb [hunk 3]
```diff
@@ -87,21 +113,90 @@ def participant_join_url(user, return_to = "http://www.instructure.com")
...
+    response = { :published => "false", :recording_formats => [] }
+    if publish=="true"
+      recording = nil
+      begin
+        response_recordings = send_request(:getRecordings, {
+        :meetingID => conference_key
+        })
+        recording = response_recordings[:recordings].find{ |r| r[:recordID]==recording_id }
+        sleep(10) if recording[:published]=="false"
+      end while recording.nil? || recording[:published]=="false"
```
Comment (simonista) [inline app/models/big_blue_button_conference.rb]: This sleep + loop happens in the context of a synchronous request, which isn't going to work, it'll tie up a passenger process for too long.  This probably needs to happen in a background job.

File: public/javascripts/big_blue_button.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
+define([
+  'i18n!plugins',
+  'jquery' /* $ */,
+  'jquery.instructure_misc_plugins' /* showIf */
+], function(I18n, $) {
+  var LEGEND_DEFAULT="Instructor can use Start/Stop Record button in sessions configured for recording."
```
Comment (jenseng) [inline public/javascripts/big_blue_button.js]: these strings all need to be internationalized

File: app/models/big_blue_button_conference.rb (hunk 3 above)
Comment (jfederico) [inline app/models/big_blue_button_conference.rb]: Wow, I didn't see that. I can't understand the intention of that, and I don't have the guy who could explain it to me :). Anyways, it is gone now

**Comment by jfederico** [comment]
Got the three changes requested after the last review. @claydiffrient I'll fix an issue in the js I just found and squash all this commits in one by forcing a push to the branch. Although, I think all the history of the reviews will be lost.

**Comment by claydiffrient** [comment]
@jfederico Nah, they'll just show as outdated I think.

**Comment by jfederico** [comment]
@claydiffrient PR has been updated. There is now a single commit. It includes the changes requested and some other updates to the conference UI. The buttons Publish/Unpublish and Delete were not working as expected. Now they are.

You can take it from here back to the code review. Let me know if you find anything else, I'll be more responsive this time (I hope ;))

I'll sketch a test plan for the people in QA. I'll be long though. Because there are several edge case scenarios that were covered.

**Comment by omarkhan** [comment]
Thanks @jfederico. I'm still seeing lots of lint failures; can you run `script/rlint` and `script/eslint` again and fix the issues they come up with?

**Comment by jfederico** [comment]
@omarkhan just came back from a one week trip, I'' be working on that. Although I only see [info] messages with rlint and [warnings] with eslint. Is there anything in particular you would like me to clean?

The other thing is that some of those warnings were there before the changes,  so the question is, how clean the code needs to be in order to be "acceptable"?

I just noticed we have again conflicts that need to be solved. I'll work on that too, but that brings another topic to my mind. Is there anything else we can do to help to accelerate the merge?

**Comment by omarkhan** [comment]
For the code to pass our internal review you'll need to fix the rlint and eslint warnings (just for your changes, don't worry about existing code). Once that's taken care of, I'll try to push this through as quickly as I can. Thanks for your patience!

**Comment by jfederico** [comment]
@omarkhan I am going to need some help man. How am I supposed to clear let say this one:

`[info] app/models/big_blue_button_conference.rb:264 => [rubocop] Prefer $ERROR_INFO from the stdlib 'English' module (don't forget to require it) over $!.

     logger.error "big blue button unhandled exception #{$!}"
                                                         ^^
`
This is one of the pieces of code we did not change, but still is not letting pass the check (as the file has been edited). I can see that it is the way things were done in other files.

I also fixed some complains about identation etc, on code that was there before, but I noticed it is shown by rlint only if the file is modified. You can add an space to wimba for instance and you'll see how the complains pop up.

Is there some way to ignore the old ones?

**Comment by jfederico** [comment]
Another clear example is this one:

This is the original code:
``
This is the code after the refactoring

**Comment by jfederico** [comment]
Hey @omarkhan I think (and hope) it looks good now. I found and fixed the xsslint issue. It wasn't that the string was not encoded but that it may be not encoded exactly where the script was expecting to find it.

I ended squashing and putting all in one commit as it was before. Forced push after that.

Anyways, please let me know the results of the review. Thanks

**Comment by simonista** [comment]
We have pulled this code in to a development branch and will be working on it internally.  Thanks for the pull request!
